### DISCUSSION TITLE
TRVZB 1.2 - External Temperature Sensor binding

### DISCUSSION THREAD

**Discussion by AliveDevil** [Koenkk/zigbee2mqtt#24487] [Devices]
With Koenkk/zigbee-ota#579 available now (I've manually updated one of my thermostats with this firmware) I'd like to support the implementation of temperature sensor to thermostat external temperature binding.
However, I have no clue where to look at - anything I can do from Zigbee2MQTT that can help in identifiying how this is implemented?

**Comment by jamesonuk** [comment]
I saw the update and have applied it to one TRV to check it out.
The change log is https://forum.ewelink.cc/t/trvzb-firmware-changelog/74377/
of which 
> Support setting external sensor, allowing SNZB-02, SNZB-02D, and SNZB-02P to be used as a reference temperature for the TRVZB.
looks interesting and what you are referring to.   Apparently is called "Temp. Display from External Sensor"  in the EweLink app.

However reading that it mentions explicit sensors and hubs make me think this is not a case of Zigbee binding.   What is odd though is if it is all done in the hub then why the firmware update on the device to support it....   There is a `customSonoffTrvzb` custom cluster which I guess is likely to be used but there is no equivalent output cluster on the sensors mentioned (that said the sensors aren't setup for OTA via Z2M)

**Reply by AliveDevil** [reply]
FWIW, this might just be a missing cluster attribute in the zigbee-herdsman-converters repo for the TRVs and climate sensors.
But as I don’t have the hardware to sniff the raw Zigbee transmissions, I thought there could be a software command that just dumps the Zigbee cluster definitions into a Json format, that can be further analyzed (i.e. an unknown reported field or something).

Is this correct?
Output Cluster: The Zigbee network exports data to the device.
Input Cluster: The Zigbee device exports data into the network.
So for a binding to work, the temperature sensors measurement input is transferred to the TRV temperature measurement output - assuming this works autonomously.

**Reply by jamesonuk** [reply]
I think the cluster tab is not part of the converter definition (IIRC this appears even one a new device without a converter and where they aren't known they appear as the numeric ID).   Whether it needs to re-interview to set it up again I don't know. (Just trying that now...)

The ZCL docs talk about client and server clusters which map to output and input clients in Z2M.    I have some switches where the output `genOnOff` cluster is bound to the `genOnOff` cluster of some lights  which is the correct way round to do things.
So it would need to be the temp sensor with the output cluster and they have not had a firmware update (in Z2M anyway)

**Reply by AliveDevil** [reply]
Gotcha. Yeah, interview didn’t do anything, and repairing didn’t have an effect either.
Could there be a firmware update for the 02P, that provides this cluster though, only available on ZBBridge?

**Reply by AliveDevil** [reply]
So, [here](https://redirect.github.com/sprut/Hub/issues/2841) there are several attributes not listed in the Z2M Clusters-list, which would contradict "cluster tab is not part of the converter definition"
E.g. 001A_RemoteSensing, 0020_StartOfWeek.

Wondering whether 0000_LocalTemperature on these TRVs is settable afterall (maybe with RemoteSensing enabled?).

**Reply by AliveDevil** [reply]
Hmm … maybe it’s really just that they are updating the local temperature offset all the time.

https://forum.ewelink.cc/t/72122/20

**Comment by AliveDevil** [comment]
Got a reply to a technical support e-mail:

> ​For the TRVZB's temp. display from the external sensor option, it only can work with ZBBridge-P now.
> […]
> If ZBBridge-P is offline, SNZB-02P disconnects with ZBBridge-P, and TRVZB will lose the detected temperature from SNZB-02P, it will use TRVZB's built-in temperature sensor to measure.

<details>
<summary>Request</summary>

> I'd like to know how I can get the external temperature sensing functionality of the ZBBridge-P for TRVZBs firmware 1.2.1 with SNZB-02P into my own Zigbee network, using Zigbee2MQTT & Home-Assistant. I have several TRVZBs deployed, but they don't appear to have the ability to be bound through Zigbee with the temperature sensors in the same network.
> 
> Is the external temperature sensing implemented by updating the local temperature offset, or does the TRVZB firmware 1.2.1 have a new bindable cluster/endpoint, that the temperature sensors can be bound to?
> 
> In case of the ZBBridge-P, if the ZBBridge-P is offline, but there are other Mains routers on the Zigbee network, keeping the network alive, do the TRVZBs continue to fetch the termperature sensors, or do these fall back to measuring the environment temperature from the built-in temperature sensor?

</details>

Not sure how this could've been implemented though.
Anyone here, who can dump the Zigbee cluster information on a 1.2.1 TRV, or has Zigbee2MQTT debug information, so one can retrieve the internal

**Comment by AliveDevil** [comment]
For reference, this is my crude automation template, which basically does the same thing as the ZBBridge Pro, assuming it only updates the temperature calibration - which it absolutely might do.

<details>
<summary>Office External Sensor</summary>

``` yaml
alias: Office Externer Sensor
description: ""
triggers:
  - trigger: state
    entity_id:
      - sensor.office_temperature
  - trigger: state
    entity_id:
      - climate.office_thermostat
    attribute: current_temperature
conditions:
  - condition: template
    value_template: "{{ has_value('sensor.office_temperature') }}"
  - condition: template
    value_template: "{{ has_value('number.office_thermostat_local_temperature_calibration') }}"
  - condition: template
    value_template: "{{ state_attr('climate.office_thermostat', 'current_temperature') != None }}"
actions:
  - action: number.set_value
    metadata: {}
    data:
      value: >-
        {{ (state_attr('climate.office_thermostat', 'current_temperature')|float -
        states('number.office_thermostat_local_temperature_calibration')|float) -
        states('sensor.office_temperature')|float }}
    target:
      entity_id: number.office_thermostat_local_temperature_calibration
mode: single
```
</details>

This assumes, that in Zigbee2MQTT the Reporting of the external sensor is set to "Min Interval: 300, Max Interval: 3600, Min Change: 10" otherwise it won't have the resolution to monitor small temperature changes well.

**Comment by AliveDevil** [comment]
For now at least, there is probably no way other than sending opening/closing valve degrees to the TRVs:

> When the connection is steady, the opening/closing of the valve will be judged by the external sensor's detected temperature. If the external sensor disconnects from the gateway, it will use the TRVZB's temperature measurement automatically, you don't need to operate it.

**Comment by bojanpotocnik** [comment]
> If the external sensor disconnects from the gateway, it will use the TRVZB's temperature measurement automatically, you don't need to operate it.

So if the sensor value is set by adjusting local temperature calibration, what happens if the sensor **and** valve, or just the valve disconnects from the gateway? Does the valve reset the local temperature calibration, or does it keep the possibly incorrect offset value, preventing it to operate correctly in the standalone mode?

**Reply by AliveDevil** [reply]
So, as far as I understand this, the ZBBridge-P directly implements either hysteresis or PID in some form directly controlling the dual valve (opened/closed percentage), and not updating the valves temperature offset.

Based on that: In case the sensor goes down, the ZBBridge-P just uses the latest value from the TRV temperature sensor (subject to _manual_ offsetting).
If the TRV looses connection, that one will just continue to use its temperature sensor and use the last set opening/closing percentage (this could very well be completely closed, when heating through e.g. external stoves).
Basically, the ZBBridge-P (or any other controller) should implement their own climate thermostat control, and feed the opening and closing valve degrees directly - ignoring the heating state, local temperature and offsets entirely.

In any case, the continuous operation of the TRV by the ZBBridge-P (or when you have some PID or hysteresis algorithm in OpenHAB, NodeRed, HA) is mandatory - and in case you want to reset the valve percentages, you‘d want to reset them through Z2M interface back to 100/100.
